sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectExecutionManager.scala

```diff
@@ -77,26 +77,28 @@ private[connect] class SparkConnectExecutionManager() extends Logging {
         request.getSessionId,
         previousSessionId)
     val executeHolder = new ExecuteHolder(request, sessionHolder)
+
+    // Check if the operation already exists, either in the active execution map, or in the
+    // graveyard of tombstones of executions that have been abandoned. The latter is to prevent
+    // double executions when the client retries, thinking it never reached the server, but in
+    // fact it did, and already got removed as abandoned.
+    if (executions.putIfAbsent(executeHolder.key, executeHolder) != null) {
+      throw new SparkSQLException(
+        errorClass = "INVALID_HANDLE.OPERATION_ALREADY_EXISTS",
+        messageParameters = Map("handle" -> executeHolder.operationId))
+    } else if (getAbandonedTombstone(executeHolder.key).isDefined) {
+      executions.remove(executeHolder.key)
+      throw new SparkSQLException(
+        errorClass = "INVALID_HANDLE.OPERATION_ABANDONED",
+        messageParameters = Map("handle" -> executeHolder.operationId))
+    }
+
+    sessionHolder.addExecuteHolder(executeHolder)
+
     executionsLock.synchronized {
-      // Check if the operation already exists, both in active executions, and in the graveyard
-      // of tombstones of executions that have been abandoned.
-      // The latter is to prevent double execution when a client retries execution, thinking it
-      // never reached the server, but in fact it did, and already got removed as abandoned.
-      if (executions.get(executeHolder.key).isDefined) {
-        throw new SparkSQLException(
-          errorClass = "INVALID_HANDLE.OPERATION_ALREADY_EXISTS",
-          messageParameters = Map("handle" -> executeHolder.operationId))
-      }
-      if (getAbandonedTombstone(executeHolder.key).isDefined) {
-        throw new SparkSQLException(
-          errorClass = "INVALID_HANDLE.OPERATION_ABANDONED",
-          messageParameters = Map("handle" -> executeHolder.operationId))
-      }
-      sessionHolder.addExecuteHolder(executeHolder)
-      executions.put(executeHolder.key, executeHolder)
```

Comment:
> But, is it now possible in some operations that an execution will be retrieved from the global map, validating that it exists, but then it will assume that it's present in the session map but race before it's added there?

As far as I looked, this looks fine, I don't see any places that first look for an operation in the global map, and then again look for it in the session map.
It would probably be a simplification to get rid of the executions in SessionHolder alltogether, and depend on the global map as a source of truth everywhere... The reason it is how it is is that before the ExecutionManager existed, the list of executions was only inside the sessions and there was no global list... Since ExecutionManager added a global utility to deal with it, it would be cleaner to get rid of the executions inside session altogether... they seem to be used in very few places anyway, but I didn't completely remove them...

Reply:
Yes, removing executions from SessionHolder will make the code simpler. My concern is, `interrupt*` will have to iterate over ExecutionManager.executions; if there are many sessions, `interrupt*` methods become inefficient.
-> Created a separate ticket for this: https://issues.apache.org/jira/browse/SPARK-49585 we can continue discuss about it in the ticket.

Agreed that there is no operation that reads ExecutionManager.sessions first assuming that there is an associated entry in SessionHolder.executions.